File: problems/1301-1400/1381_design-a-stack-with-increment-operation/solution.go
package main

type CustomStack struct {
	values  []int64
	pending []int64
	maxSize int
}

func NewCustomStackTyped(maxSize int) *CustomStack {
	return &CustomStack{maxSize: maxSize}
}

func (design *CustomStack) push(x int) {
	if len(design.values) < design.maxSize {
		design.values = append(design.values, int64(x))
		design.pending = append(design.pending, 0)
	}
}

func (design *CustomStack) pop() int {
	if len(design.values) == 0 {
		return -1
	}
	increment := design.pending[len(design.pending)-1]
	design.pending = design.pending[:len(design.pending)-1]
	if len(design.pending) > 0 {
		design.pending[len(design.pending)-1] += increment
	}
	value := design.values[len(design.values)-1] + increment
	design.values = design.values[:len(design.values)-1]
	return int(value)
}

func (design *CustomStack) increment(k int, val int) {
	limit := min(k, len(design.values))
	if limit > 0 {
		design.pending[limit-1] += int64(val)
	}
}

File: problems/1301-1400/1381_design-a-stack-with-increment-operation/solution.rs
pub struct CustomStack {
    values: Vec<i64>,
    pending: Vec<i64>,
    max_size: usize,
}

impl CustomStack {
    pub fn new(max_size: i32) -> Self {
        CustomStack {
            values: Vec::new(),
            pending: Vec::new(),
            max_size: max_size as usize,
        }
    }

    pub fn push(&mut self, x: i32) {
        if self.values.len() < self.max_size {
            self.values.push(x as i64);
            self.pending.push(0);
        }
    }

    pub fn pop(&mut self) -> i32 {
        let value = match self.values.pop() {
            None => return -1,
            Some(value) => value,
        };
        let increment = self.pending.pop().unwrap();
        if let Some(deepest) = self.pending.last_mut() {
            *deepest += increment;
        }
        (value + increment) as i32
    }

    pub fn increment(&mut self, k: i32, val: i32) {
        let limit = (k as i64).min(self.values.len() as i64);
        if limit > 0 {
            self.pending[(limit - 1) as usize] += val as i64;
        }
    }
}
